src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java

```diff
@@ -27,165 +27,264 @@
 import io.fabric8.openshift.api.model.BuildStrategy;
 import io.fabric8.openshift.api.model.GitBuildSource;
 import io.fabric8.openshift.api.model.JenkinsPipelineBuildStrategy;
+
+import jenkins.branch.Branch;
 import org.apache.commons.lang.StringUtils;
 import org.eclipse.jgit.transport.RemoteConfig;
 import org.eclipse.jgit.transport.URIish;
 import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
 import org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition;
 import org.jenkinsci.plugins.workflow.flow.FlowDefinition;
 import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.multibranch.BranchJobProperty;
+import org.jenkinsci.plugins.workflow.multibranch.WorkflowBranchProjectFactory;
 
 import java.io.File;
 import java.io.IOException;
 import java.util.Collections;
 import java.util.List;
+import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import static io.fabric8.jenkins.openshiftsync.Annotations.GENERATED_BY;
+import static io.fabric8.jenkins.openshiftsync.Annotations.GROOVY_SANDBOX;
 import static io.fabric8.jenkins.openshiftsync.CredentialsUtils.updateSourceCredentials;
+import static io.fabric8.jenkins.openshiftsync.EnvironmentVariables.OPENSHIFT_ALLOW_GROOVY_SANDBOX_DISABLED;
+import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.addAnnotation;
+import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.getAnnotation;
+import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.removeAnnotation;
 import static org.apache.commons.lang.StringUtils.isEmpty;
 import static org.apache.commons.lang.StringUtils.isNotBlank;
 
 public class BuildConfigToJobMapper {
-  public static final String JENKINS_PIPELINE_BUILD_STRATEGY = "JenkinsPipeline";
-  public static final String DEFAULT_JENKINS_FILEPATH = "Jenkinsfile";
-  private static final Logger LOGGER = Logger.getLogger(BuildConfigToJobMapper.class.getName());
+	public static final String JENKINS_PIPELINE_BUILD_STRATEGY = "JenkinsPipeline";
+	public static final String DEFAULT_JENKINS_FILEPATH = "Jenkinsfile";
+	private static final Logger LOGGER = Logger.getLogger(BuildConfigToJobMapper.class.getName());
 
-  public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOException {
-    if (!OpenShiftUtils.isJenkinsBuildConfig(bc)) {
-      return null;
-    }
+	public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOException {
```

Comment:
the changes here broke our openshift regression testing when leveraging a jenkinsfilePath.  I got this branch working again by changing this method as follows:

```
    public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc)
            throws IOException {
        if (!OpenShiftUtils.isPipelineStrategyBuildConfig(bc)) {
            return null;
        }

        // if the BC was created by jenkins then default to the SCMBinder
        String generatedBy = getAnnotation(bc, GENERATED_BY);
        if (generatedBy != null && generatedBy.equals("jenkins")) {
            return new ExposeSCMBinderHack().createDefinition();
        }

        BuildConfigSpec spec = bc.getSpec();
        BuildSource source = null;
        String jenkinsfile = null;
        String jenkinsfilePath = null;
        if (spec != null) {
            source = spec.getSource();
            BuildStrategy strategy = spec.getStrategy();
            if (strategy != null) {
                JenkinsPipelineBuildStrategy jenkinsPipelineStrategy = strategy
                        .getJenkinsPipelineStrategy();
                if (jenkinsPipelineStrategy != null) {
                    jenkinsfile = jenkinsPipelineStrategy.getJenkinsfile();
                    jenkinsfilePath = jenkinsPipelineStrategy
                            .getJenkinsfilePath();
                }
            }
        }
        
        // note: openshift validation will ensure that both jenkinsfile
        // and jenkinsfilePath are not set, so this code does not need
        // to worry about that; been that way since the inception of
        // pipeline build strategy
                
        if (jenkinsfile != null && jenkinsfile.trim().length() > 0) {
            boolean sandbox = true;
            String sandboxFlag = getAnnotation(bc, GROOVY_SANDBOX);
            if (sandboxFlag != null && sandboxFlag.equalsIgnoreCase("false")) {
                // lets only allow this if this env var is specified
                String allowed = System
                        .getenv(OPENSHIFT_ALLOW_GROOVY_SANDBOX_DISABLED);
                if (allowed != null && allowed.equalsIgnoreCase("true")) {
                    sandbox = false;
                }
            }
            return new CpsFlowDefinition(jenkinsfile, sandbox);
        
        }
        
        if (source != null && source.getGit() != null
                && source.getGit().getUri() != null) {
            if (jenkinsfilePath == null || 
                jenkinsfilePath.trim().length() == 0) {
                jenkinsfilePath = DEFAULT_JENKINS_FILEPATH;
            }
            if (!isEmpty(source.getContextDir())) {
                jenkinsfilePath = new File(source.getContextDir(),
                        jenkinsfilePath).getPath();
            }
            GitBuildSource gitSource = source.getGit();
            String branchRef = gitSource.getRef();
            List<BranchSpec> branchSpecs = Collections.emptyList();
            if (isNotBlank(branchRef)) {
                branchSpecs = Collections.singletonList(new BranchSpec(
                        branchRef));
            }
            String credentialsId = updateSourceCredentials(bc);
            // if credentialsID is null, go with an SCM where anonymous has
            // to be sufficient
            GitSCM scm = new GitSCM(
                    Collections.singletonList(new UserRemoteConfig(
                            gitSource.getUri(), null, null, credentialsId)),
                    branchSpecs, false, Collections
                            .<SubmoduleConfig> emptyList(), null, null,
                    Collections.<GitSCMExtension> emptyList());
            return new CpsScmFlowDefinition(scm, jenkinsfilePath);
        }
        
        LOGGER.warning("BuildConfig does not contain source repository information - "
                + "cannot map BuildConfig to Jenkins job");
        return null;
    }

```